Document 406 Not Acceptable (#2)
chris-wood/draft-unreliable-ohttp: draft-wood-ohai-unreliable-ohttp.md

Encapsulated Request as described in {{OHTTP}}. This Encapsulated Request
is included as the content of a POST request to the Oblivious Relay Resource.
Importantly, this request MUST include the "message/ohttp-ack" Media Type
in the Accept header (see {{iana-ack}}). The Client receives a `202 Accepted`
response with content type "message/ohttp-ack" and an empty body upon successful
transmission of the request. If the Accept header also allows the normal
"message/bhttp" content type, the client has left the choice of reliable or

> unreliable delivery up to the Relay and Gateway Resources, which may
> return "200 OK" along with an Encapsulated Body instead.
> Any other response is considered invalid.

Reviewer: ```suggestion
unreliable delivery up to the Relay and Gateway Resources.
```
Let's let the main spec (covering reliable mode) deal with what responses are invalid and not repeat them here.

Upon receipt of an unreliable OHTTP request from the Client, the Oblivious
Relay Resource MUST reply with a `202 Accepted` response with the "message/ohttp-ack"

> content type to the Client. It SHOULD buffer the request to be sent to the Oblivious
> Gateway Resource at some point in the future, but MAY chose to forward it immediately.

Reviewer: ```suggestion
content type to the Client. It can buffer the request to be sent to the Oblivious
Gateway Resource at some point in the future, or choose to forward it immediately.
```

I don't think we should require buffering here, since that's an implementation-specific decision.

Author: Ok

code in the response from the Relay Resource, this means that the OHTTP
configuration for that request only supports unreliable OHTTP. The client
MAY then retry, this time allowing unreliable delivery.

## Relay Considerations

Unreliable OHTTP allows the Oblivious Relay Resource to buffer Encapsulated Requests
for future transmission to an Oblivious Gateway Resource. The relay can choose to
buffer Client requests until a sufficiently large number of requests is reached
and then send all requests in a single batch to the gateway. Additionally, the relay
can shuffle requests before forwarding them to the gateway to further obscure
the request timing.

The choice of minimum buffer size is an implementation detail. Relays should take care
to not introduce too much delay between when a request was received and when it is
forwarded. Such delay may cause the gateway to drop the request due to a variety
of reasons, e.g., because the gateway configuration changed or rotated or the decapsulated
request became too old.

A Relay which supports unreliable OHTTP may be configured to enforce
unreliable delivery. In such cases, if the Relay receives a normal

> OHTTP request, i.e. without an Accept header matching "message/ohttp-ack",

Author: I see you're a "more commas" person. :)

any delay in processing a request can increase the likelihood that the request
is dropped or discarded due to it being stale.

If an Oblivious Gateway Resource requires unreliable delivery for a
request, by implementation constraint or policy, it SHOULD respond

> with `406 Not Acceptable` to any requests which require return of an

Reviewer: ```suggestion
with `406 Not Acceptable` to any requests which require reliable return of an
```

Encapsulated Response. The Relay Resource will forward this to the

> Client, which can fail or retry with unreliable delivery according to

Reviewer: ```suggestion
Client, and the Client can then fail or retry with unreliable delivery according to
```

its own requirements.

> Note that while it's possible for the Gateway to decide whether to

Reviewer: This is interesting =) Given the privacy implications here, I would recommend we remove this text and file an issue to consider adding it back.

Author: It might make more sense in the Security section, but only after we add the `Accept: message/*` bit. I will move.

Author: https://github.com/chris-wood/draft-unreliable-ohttp/issues/3
